File: local-server.py
#!/usr/bin/env python3
import json
import mimetypes
import os
import re
from datetime import datetime, timezone
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

ROOT = Path(__file__).resolve().parent
PORT = int(os.environ.get("PORT", "5175"))
HOST = "127.0.0.1"

SETTING_KEYS = [
    "bendScale",
    "shearScale",
    "lightScale",
    "chromaScale",
    "speedScale",
    "coverage",
    "baseBrightness",
    "gamma",
    "shadowScale",
]
BACKGROUND_SETTING_KEYS = [
    "backgroundBlur",
    "backgroundBrightness",
    "backgroundSaturation",
]


def format_number(key, value):
    number = float(value)
    if key == "baseBrightness" or number.is_integer():
        return str(int(number))
    return f"{number:.2f}".rstrip("0").rstrip(".")


def clean_settings(settings, keys):
    return {
        key: float(settings.get(key, 0))
        for key in keys
    }


def save_ripple_settings(settings):
    ripple_source = settings.get("ripple", settings)
    background_source = settings.get("background", {})
    clean = {
        key: float(ripple_source.get(key, 0))
        for key in SETTING_KEYS
    }
    clean_background = {
        key: float(background_source.get(key, 0))
        for key in BACKGROUND_SETTING_KEYS
    }
    saved_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    version = str(int(datetime.now().timestamp() * 1000))

    app_path = ROOT / "app.js"
    app = app_path.read_text(encoding="utf-8")
    settings_block = "\n".join(
        f"  {key}: {format_number(key, clean[key])},"
        for key in SETTING_KEYS
    )
    app = re.sub(
        r"const defaultRippleSettings = \{[\s\S]*?\};",
        f"const defaultRippleSettings = {{\n{settings_block}\n}};",
        app,
    )
    background_block = "\n".join(
        f"  {key}: {format_number(key, clean_background[key])},"
        for key in BACKGROUND_SETTING_KEYS
    )
    app = re.sub(
        r"const defaultBackgroundSettings = \{[\s\S]*?\};",
        f"const defaultBackgroundSettings = {{\n{background_block}\n}};",
        app,
    )
    app_path.write_text(app, encoding="utf-8")

    html_path = ROOT / "index.html"
    html = html_path.read_text(encoding="utf-8")
    html = re.sub(r"styles\.css\?v=[^\"]+", f"styles.css?v={version}", html)
    html = re.sub(r"app\.js\?v=[^\"]+", f"app.js?v={version}", html)
    if 'name="ripple-settings-saved-at"' in html:
        html = re.sub(
            r'<meta name="ripple-settings-saved-at" content="[^"]*" />',
            f'<meta name="ripple-settings-saved-at" content="{saved_at}" />',
            html,
        )
    else:
        html = html.replace(
            "</head>",
            f'    <meta name="ripple-settings-saved-at" content="{saved_at}" />\n  </head>',
        )
    html_path.write_text(html, encoding="utf-8")

    css_path = ROOT / "styles.css"
    css = css_path.read_text(encoding="utf-8")
    marker = f"/* ripple-settings-saved-at: {saved_at} */"
    if css.startswith("/* ripple-settings-saved-at:"):
        css = re.sub(r"^/\* ripple-settings-saved-at: .* \*/\n", marker + "\n", css)
    else:
        css = marker + "\n" + css
    css = re.sub(r"--background-blur: [^;]+;", f"--background-blur: {format_number('backgroundBlur', clean_background['backgroundBlur'])}px;", css)
    css = re.sub(r"--background-brightness: [^;]+;", f"--background-brightness: {format_number('backgroundBrightness', clean_background['backgroundBrightness'])};", css)
    css = re.sub(r"--background-saturation: [^;]+;", f"--background-saturation: {format_number('backgroundSaturation', clean_background['backgroundSaturation'])};", css)
    css_path.write_text(css, encoding="utf-8")


def save_cards(cards):
    if not isinstance(cards, list):
        raise ValueError("cards payload must be a list")

    cleaned = []
    for index, card in enumerate(cards):
        if not isinstance(card, dict):
            raise ValueError(f"card {index} must be an object")
        pose = card.get("pose", {})
        if not isinstance(pose, dict):
            pose = {}
        cleaned.append({
            "id": str(card.get("id") or f"card-{index + 1}"),
            "label": str(card.get("label", "")),
            "sub-label": str(card.get("sub-label", card.get("subLabel", ""))),
            "label-fill": str(card.get("label-fill", card.get("labelFill", ""))),
            "image": str(card.get("image") or ""),
            "pose": pose,
        })

    cards_path = ROOT / "data" / "cards.json"
    cards_path.parent.mkdir(parents=True, exist_ok=True)
    cards_path.write_text(
        json.dumps(cleaned, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )

    version = str(int(datetime.now().timestamp() * 1000))
    saved_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    html_path = ROOT / "index.html"
    html = html_path.read_text(encoding="utf-8")
    html = re.sub(r"styles\.css\?v=[^\"]+", f"styles.css?v={version}", html)
    html = re.sub(r"app\.js\?v=[^\"]+", f"app.js?v={version}", html)
    if 'name="cards-saved-at"' in html:
        html = re.sub(
            r'<meta name="cards-saved-at" content="[^"]*" />',
            f'<meta name="cards-saved-at" content="{saved_at}" />',
            html,
        )
    else:
        html = html.replace(
            "</head>",
            f'    <meta name="cards-saved-at" content="{saved_at}" />\n  </head>',
        )
    html_path.write_text(html, encoding="utf-8")


def save_focus_layout(layout):
    if not isinstance(layout, dict):
        raise ValueError("focus layout payload must be an object")

    slot_order = ["front", "left", "right", "back"]
    fallback_labels = {
        "front": "前景",
        "left": "左侧",
        "right": "右侧",
        "back": "后方",
    }
    numeric_keys = ["x", "y", "z", "r", "rx", "ry", "s", "blur", "fade", "sat"]
    cleaned = {}
    for slot in slot_order:
        source = layout.get(slot, {})
        if not isinstance(source, dict):
            source = {}
        cleaned[slot] = {"label": str(source.get("label") or fallback_labels[slot])}
        for key in numeric_keys:
            cleaned[slot][key] = float(source.get(key, 0))

    layout_path = ROOT / "data" / "focus-layout.json"
    layout_path.parent.mkdir(parents=True, exist_ok=True)
    layout_path.write_text(
        json.dumps(cleaned, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )

    version = str(int(datetime.now().timestamp() * 1000))
    saved_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    html_path = ROOT / "index.html"
    html = html_path.read_text(encoding="utf-8")
    html = re.sub(r"styles\.css\?v=[^\"]+", f"styles.css?v={version}", html)
    html = re.sub(r"app\.js\?v=[^\"]+", f"app.js?v={version}", html)
    if 'name="focus-layout-saved-at"' in html:
        html = re.sub(
            r'<meta name="focus-layout-saved-at" content="[^"]*" />',
            f'<meta name="focus-layout-saved-at" content="{saved_at}" />',
            html,
        )
    else:
        html = html.replace(
            "</head>",
            f'    <meta name="focus-layout-saved-at" content="{saved_at}" />\n  </head>',
        )
    html_path.write_text(html, encoding="utf-8")


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def end_headers(self):
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def do_POST(self):
        if self.path not in {"/api/ripple-settings", "/api/cards", "/api/focus-layout"}:
            self.send_response(404)
            self.end_headers()
            return

        length = int(self.headers.get("Content-Length", "0"))
        payload = self.rfile.read(length).decode("utf-8")
        try:
            data = json.loads(payload or "{}")
            if self.path == "/api/ripple-settings":
                save_ripple_settings(data)
            elif self.path == "/api/cards":
                save_cards(data)
            else:
                save_focus_layout(data)
            body = b'{"ok":true}'
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
        except Exception as error:
            body = str(error).encode("utf-8")
            self.send_response(500)
            self.send_header("Content-Type", "text/plain; charset=utf-8")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)


if __name__ == "__main__":
    mimetypes.add_type("text/javascript; charset=utf-8", ".js")
    server = ThreadingHTTPServer((HOST, PORT), Handler)
    print(f"One-Hop Motion preview: http://{HOST}:{PORT}")
    server.serve_forever()

File: local-server.mjs
import { createServer } from "node:http";
import { readFile, writeFile } from "node:fs/promises";
import { extname, join, normalize } from "node:path";

const root = new URL(".", import.meta.url).pathname;
const port = Number(process.env.PORT || 5174);
const host = "127.0.0.1";

const mimeTypes = {
  ".html": "text/html; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".png": "image/png",
};

const settingKeys = [
  "bendScale",
  "shearScale",
  "lightScale",
  "chromaScale",
  "speedScale",
  "coverage",
  "baseBrightness",
  "gamma",
  "shadowScale",
];
const backgroundSettingKeys = [
  "backgroundBlur",
  "backgroundBrightness",
  "backgroundSaturation",
];

function formatSettings(settings) {
  return settingKeys
    .map((key) => `  ${key}: ${Number(settings[key]).toFixed(key === "baseBrightness" ? 0 : 2).replace(/\.00$/, "")},`)
    .join("\n");
}

function sanitizeSettings(settings) {
  return Object.fromEntries(
    settingKeys.map((key) => [key, Number.isFinite(Number(settings[key])) ? Number(settings[key]) : 0])
  );
}

async function saveRippleSettings(settings) {
  const clean = sanitizeSettings(settings.ripple ?? settings);
  const cleanBackground = Object.fromEntries(
    backgroundSettingKeys.map((key) => [key, Number.isFinite(Number(settings.background?.[key])) ? Number(settings.background[key]) : 0])
  );
  const savedAt = new Date().toISOString();

  const appPath = join(root, "app.js");
  let app = await readFile(appPath, "utf8");
  app = app.replace(
    /const defaultRippleSettings = \{[\s\S]*?\};/,
    `const defaultRippleSettings = {\n${formatSettings(clean)}\n};`
  );
  app = app.replace(
    /const defaultBackgroundSettings = \{[\s\S]*?\};/,
    `const defaultBackgroundSettings = {\n${backgroundSettingKeys
      .map((key) => `  ${key}: ${Number(cleanBackground[key]).toFixed(key === "backgroundBlur" ? 1 : 2).replace(/\.0$/, "").replace(/\.00$/, "")},`)
      .join("\n")}\n};`
  );
  await writeFile(appPath, app);

  const htmlPath = join(root, "index.html");
  let html = await readFile(htmlPath, "utf8");
  html = html.replace(/styles\.css\?v=[^"]+/, `styles.css?v=${Date.now()}`);
  html = html.replace(/app\.js\?v=[^"]+/, `app.js?v=${Date.now()}`);
  html = html.includes('name="ripple-settings-saved-at"')
    ? html.replace(/<meta name="ripple-settings-saved-at" content="[^"]*" \/>/, `<meta name="ripple-settings-saved-at" content="${savedAt}" />`)
    : html.replace("</head>", `    <meta name="ripple-settings-saved-at" content="${savedAt}" />\n  </head>`);
  await writeFile(htmlPath, html);

  const cssPath = join(root, "styles.css");
  let css = await readFile(cssPath, "utf8");
  const marker = `/* ripple-settings-saved-at: ${savedAt} */`;
  css = css.startsWith("/* ripple-settings-saved-at:")
    ? css.replace(/^\/\* ripple-settings-saved-at: .* \*\/\n/, `${marker}\n`)
    : `${marker}\n${css}`;
  css = css
    .replace(/--background-blur: [^;]+;/, `--background-blur: ${cleanBackground.backgroundBlur}px;`)
    .replace(/--background-brightness: [^;]+;/, `--background-brightness: ${cleanBackground.backgroundBrightness};`)
    .replace(/--background-saturation: [^;]+;/, `--background-saturation: ${cleanBackground.backgroundSaturation};`);
  await writeFile(cssPath, css);
}

function resolveRequestPath(url) {
  const requestPath = decodeURIComponent(new URL(url, `http://${host}:${port}`).pathname);
  const normalized = normalize(requestPath === "/" ? "/index.html" : requestPath).replace(/^(\.\.[/\\])+/, "");
  return join(root, normalized);
}

createServer(async (request, response) => {
  try {
    if (request.method === "POST" && request.url === "/api/ripple-settings") {
      let body = "";
      request.on("data", (chunk) => {
        body += chunk;
      });
      request.on("end", async () => {
        await saveRippleSettings(JSON.parse(body || "{}"));
        response.writeHead(200, { "Content-Type": "application/json" });
        response.end(JSON.stringify({ ok: true }));
      });
      return;
    }

    if (request.method !== "GET") {
      response.writeHead(405);
      response.end();
      return;
    }

    const filePath = resolveRequestPath(request.url);
    if (!filePath.startsWith(root)) {
      response.writeHead(403);
      response.end();
      return;
    }

    const content = await readFile(filePath);
    response.writeHead(200, {
      "Content-Type": mimeTypes[extname(filePath)] || "application/octet-stream",
      "Cache-Control": "no-store",
    });
    response.end(content);
  } catch (error) {
    response.writeHead(error.code === "ENOENT" ? 404 : 500, { "Content-Type": "text/plain; charset=utf-8" });
    response.end(error.code === "ENOENT" ? "Not found" : String(error.message || error));
  }
}).listen(port, host, () => {
  console.log(`One-Hop Motion preview: http://${host}:${port}`);
});
